### PR TITLE
Add Capsule Layer to layers (capsule.py)

### PR DESCRIPTION
**Pull request by SriRangaTarun**
I have added a Capsule Layer implementation (pure Keras), referring to [this example](https://github.com/keras-team/keras/blob/master/examples/cifar10_cnn_capsule.py) on the Keras repository. [Here](https://github.com/keras-team/keras/issues/12097) is the related issue on the Keras repository.

### CODE REVIEW

**Comment by gabrieldemarmiesse** [comment]
Do you need any help with this PR?

**Comment by SriRangaTarun** [comment]
Yes, @gabrieldemarmiesse, I guess I do need help. If possible, can you brief me on what kind of mistakes people do when adding a new layer.

**Comment by gabrieldemarmiesse** [comment]
Ususally, the workflow is like this:
* Implement the layer in a file (I think you've already done this)
* Implement tests for this layer, put them with the other layer tests, you can see other tests to find out what is the syntax to use and how unit tests are done.
* Ensure that your code is PEP8 complient.
* At this point, the Travis CI build should pass. If the Travis CI build failed (like now) it's either a test failure (your layer is incorrect), either a PEP8 error. You can see the logs of travis explaining why it failed by clicking `Details` next to the travis logo at the bottom of the pull request.
* Ask for a review.

If you don't do at least the first three steps, it's hard for me to help you because I won't see any failure from Travis CI, so I cannot see the bugs.

**Comment by SriRangaTarun** [comment]
@gabrieldemarmiesse. Thanks for replying. Everything seems to be working in the Travis CI Build, except the PEP8 formatting. What do people on Keras generally do to make sure their code is PEP8 complient ? I tried many python packages for formatting, but none of them worked. Is it done manually ?

**Comment by gabrieldemarmiesse** [comment]
You PR lacks unit tests. To solve the PEP8 problems, either read the travis logs, or you can also use the instructions in `CONTRIBUTING.md`.

**Comment by SriRangaTarun** [comment]
I'll add unit tests soon.

**Comment by SriRangaTarun** [comment]
@gabrieldemarmiesse, I confirm that the tests have also run successfully.

**Comment by gabrieldemarmiesse** [comment]
It seems that the travis.ci build isn't passing. Can you fix the errors? You can find the logs by clicking "Details" next to the travis CI logo.

**Comment by SriRangaTarun** [comment]
The build started failing only after I added the tests.

**Comment by SriRangaTarun** [comment]
@gabrieldemarmiesse, it looks like the Travis CI Build has passed and everything is working fine. Please inform me if I need to add better tests. I don't know if this is relevant, but this same implementation works as expected for me in the current project I'm doing.

**Comment by SriRangaTarun** [comment]
@gabrieldemarmiesse, it looks like the Travis CI Build has passed and everything is working fine. It turns out, the K.batch_dot was calculating the elementwise products correctly, but it wasn't adding up the products. So, I just added K.sum. Please inform me if I need to add better tests. I don't know if this is relevant, but this same implementation works as expected for me in the current project I'm doing.

**Comment by SriRangaTarun** [comment]
@gabrieldemarmiesse, does everything look fine to you ? If not, kindly suggest any changes.

File: keras_contrib/layers/capsule.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+from keras.utils import get_custom_objects
+
+
+def squash(x, axis=-1):
+    s_squared_norm = K.sum(K.square(x), axis, keepdims=True) + K.epsilon()
+    scale = K.sqrt(s_squared_norm) / (0.5 + s_squared_norm)
+    return scale * x
+
+
+def softmax(x, axis=-1):
```
[the remark this answers is not in the document]
Comment (SriRangaTarun) [reply keras_contrib/layers/capsule.py]: Yes, sure. I'll use K.softmax. It dosen't really make a difference.

[the remark this answers is not in the document]
Comment (SriRangaTarun) [reply keras_contrib/layers/capsule.py]: Is the documentation fine ?

[the remark this answers is not in the document]
Comment (SriRangaTarun) [reply keras_contrib/layers/capsule.py]: @gabrieldemarmiesse ?

[the remark this answers is not in the document]
Comment (gabrieldemarmiesse) [reply keras_contrib/layers/capsule.py]: I'm quite busy at the moment. I only develop keras-contrib in my free time, so please be patient. I think I'll review this weekend.

File: keras_contrib/layers/capsule.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+        for i in range(self.routings):
+            c = K.softmax(b, 1)
+            o = K.batch_dot(c, u_hat_vecs, [2, 2])
+            if K.backend() == 'theano' or len(K.int_shape(o)) == 4:
+                o = K.sum(o, axis=1)
+            if i < self.routings - 1:
+                o = K.l2_normalize(o, -1)
+                b = K.batch_dot(o, u_hat_vecs, [2, 3])
+                if K.backend() == 'theano' or len(K.int_shape(b)) == 4:
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: Could you explain why you would need to do that with only theano? What makes it different?

File: keras_contrib/layers/capsule.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+        - [Dynamic-Routing-Between-Capsules]
+          (https://arxiv.org/pdf/1710.09829.pdf)
+        - [Keras-Examples-CIFAR10-CNN-Capsule]"""
+
+    def __init__(self,
+                 num_capsule,
+                 dim_capsule,
+                 routings=3,
+                 share_weights=True,
+                 activation='squash',
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: Default activation should be `None` (linear) to be consistent with the rest of keras. You should add in the docstring that a capsule network is mainly used a the `squash` activation. But it should not be possible to use the squash activation with just a string (to be again consistant with keras). Please show an example of how to use the squash activation.

File: keras_contrib/layers/capsule.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        self.dim_capsule = dim_capsule
+        self.routings = routings
+        self.share_weights = share_weights
+        if activation == 'squash':
+            self.activation = squash
+        else:
+            self.activation = activations.get(activation)
+
+    def build(self, input_shape):
+        super(Capsule, self).build(input_shape)
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: You shouldn't do that. just use `self.build = True` at the end of your method. See https://github.com/keras-team/keras/blob/a1397169ddf8595736c01fcea084c8e34e1a3884/keras/layers/core.py#L887

File: keras_contrib/layers/capsule.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+        input_dim_capsule = input_shape[-1]
+        if self.share_weights:
+            self.W = self.add_weight(name='capsule_kernel',
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: The weights should have an `initializer`, `regularizer` and `constrain` which can be used specified at the layer creation time. Please see https://github.com/keras-team/keras/blob/a1397169ddf8595736c01fcea084c8e34e1a3884/keras/layers/core.py#L891 as an example. The default initializer should still be `'glorot_uniform'` like you did.

File: keras_contrib/layers/capsule.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+                                     shape=(1,
+                                            input_dim_capsule,
+                                            self.num_capsule *
+                                            self.dim_capsule),
+                                     initializer='glorot_uniform',
+                                     trainable=True)
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: No need to specify the default value here.

File: keras_contrib/layers/capsule.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+        config = {'num_capsule': self.num_capsule,
+                  'dim_capsule': self.dim_capsule,
+                  'routings': self.routings,
+                  'share_weights': self.share_weights,
+                  'activation': activations.serialize(self.activation)}
+
+        base_config = super(Capsule, self).get_config()
+        return dict(list(base_config.items()) + list(config.items()))
+
+get_custom_objects().update({'Capsule': Capsule})
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: I know all other classes do this, but let's not add it to the custom objects. Users will do it manually. Otherwise we'll end up with conflicts of other python packages do the same.

**Comment by SriRangaTarun** [comment]
Thanks for reviewing the code, @gabrieldemarmiesse ! I will make the necessary changes by tomorrow or at the latest, day after tomorrow.

File: keras_contrib/layers/capsule.py (hunk 7 above)
Comment (SriRangaTarun) [reply keras_contrib/layers/capsule.py]: Removed that line.

File: keras_contrib/layers/capsule.py (hunk 6 above)
Comment (SriRangaTarun) [reply keras_contrib/layers/capsule.py]: Changed it to 'None'.

File: keras_contrib/layers/capsule.py (hunk 5 above)
Comment (SriRangaTarun) [reply keras_contrib/layers/capsule.py]: Added an initlaizer, regularizer and constraint to the weights.

File: keras_contrib/layers/capsule.py (hunk 4 above)
Comment (SriRangaTarun) [reply keras_contrib/layers/capsule.py]: Fixed it.

File: keras_contrib/layers/capsule.py (hunk 3 above)
Comment (SriRangaTarun) [reply keras_contrib/layers/capsule.py]: Changed the default activation to 'None'.  Added an example of how to use the squash function as activation in the docstring. Removed the option of using the squash activation function with a string.

File: keras_contrib/layers/capsule.py (hunk 2 above)
Comment (SriRangaTarun) [reply keras_contrib/layers/capsule.py]: I thought that the Theano backend batch_dot behaves differently than the TensorFlow one. But, it doesn't. Removed it.

**Comment by SriRangaTarun** [comment]
@gabrieldemarmiesse, it looks like the Travis CI Build Test is passing , but I have made all the changes you proposed except one : removing the line which adds Capsule to the custom objects. Whenever I try to do this, the Travis CI Build fails, saying that "The 'Capsule' class is not found". Any idea about what could be going wrong ?

**Comment by gabrieldemarmiesse** [comment]
I see where the problem comes from. We'll merge it without removing the custom object line. I'll remove it myself in another pull request later on. Thanks for your work! 

Let me a bit of time to review it again. Or maybe @RaphaelMeudec if you have the time?

File: keras_contrib/layers/capsule.py [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+       # Arguments
+           num_capsule : Number of Capsules (int)
+           dim_capsules : Dimensions of the vector output of each Capsule (int)
+           routings : Number of dynamic routings in the Capsule Layer (int)
+           share_weights : Whether to share weights between Capsules or not
+           (boolean)
+           activation : Activation function for the Capsules (str)
+           regularizer : Regularizer for the weights of the Capsules (str)
+           initializer : Initializer for the weights of the Caspules (str)
+           cnstraint : Constraint for the weights of the Capsules (str)
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: Typo. And note that it's not necessarily a string.

Comment (SriRangaTarun) [reply keras_contrib/layers/capsule.py]: Fixed the typo and removed 'str'. Included the option of adding initializers, regularizers and constraints without string.

**Comment by gabrieldemarmiesse** [comment]
@RaphaelMeudec if you want to fix this custom objects issue, see #420 which is a similar fix for optimizers.

**Comment by SriRangaTarun** [comment]
@RaphaelMeudec, I changed the long if-else statements to one-liners. But, the variable names correlate well with how the tensors are named in Hinton's original paper. For example, look at this [blog about the paper](https://medium.com/ai%C2%B3-theory-practice-business/understanding-hintons-capsule-networks-part-ii-how-capsules-work-153b6ade9f66)

**Comment by SriRangaTarun** [comment]
Or should I add more elaborate names, like 'capsule_scores' for 'c'.

**Comment by SriRangaTarun** [comment]
But, I could not find any better name than 'u_hat_vectors' for the spatially transformed inputs.

File: keras_contrib/layers/capsule.py [hunk 9]
```diff
@@ -0,0 +1,216 @@
...
+                 constraint=None,
+                 **kwargs):
+        super(Capsule, self).__init__(**kwargs)
+        self.num_capsule = num_capsule
+        self.dim_capsule = dim_capsule
+        self.routings = routings
+        self.share_weights = share_weights
+
+        self.activation = activations.get(activation) \
+            if (type(activation) == str or activation is None) else activation
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: There is no need for the if-else statement. The get function handles automatically everything.
`self.activation = activations.get(activation)` is enough.

File: keras_contrib/layers/capsule.py [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+
+        self.regularizer = regularizers.get(regularizer) \
+            if (type(regularizer) == str or regularizer is None) else regularizer
+
+        self.initializer = initializers.get(initializer) \
+            if (type(initializer) == str) else initializer
+
+        self.constraint = constraints.get(constraint) \
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: same here.

File: keras_contrib/layers/capsule.py [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+        self.routings = routings
+        self.share_weights = share_weights
+
+        self.activation = activations.get(activation) \
+            if (type(activation) == str or activation is None) else activation
+
+        self.regularizer = regularizers.get(regularizer) \
+            if (type(regularizer) == str or regularizer is None) else regularizer
+
+        self.initializer = initializers.get(initializer) \
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: same here.

File: keras_contrib/layers/capsule.py [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+        super(Capsule, self).__init__(**kwargs)
+        self.num_capsule = num_capsule
+        self.dim_capsule = dim_capsule
+        self.routings = routings
+        self.share_weights = share_weights
+
+        self.activation = activations.get(activation) \
+            if (type(activation) == str or activation is None) else activation
+
+        self.regularizer = regularizers.get(regularizer) \
```
Comment (gabrieldemarmiesse) [inline keras_contrib/layers/capsule.py]: same here.

**Comment by SriRangaTarun** [comment]
@gabrieldemarmiesse, I have removed all the necessary if-else statements. Anything else ?

**Comment by gabrieldemarmiesse** [comment]
If you merge master into your branch, you should be able to remove the line with the custom objects now.

**Comment by SriRangaTarun** [comment]
@gabrieldemarmiesse, I've merged the master into my branch and removed the custom objects line.

File: tests/keras_contrib/layers/test_capsule.py [hunk 13]
```diff
@@ -0,0 +1,51 @@
...
+
+    model = Sequential()
+    model.add(capsule.Capsule(1, 1, 1, True, activation='sigmoid'))
+
+    model.compile(loss='mse', optimizer='rmsprop')
+    init_out = model.predict(X)  # mock predict call to initialize weights
+    model.set_weights([np.zeros((1, 1, 1))])
+    out = model.predict(X)
+    assert_allclose(out, np.zeros((1, 1, 1), dtype=K.floatx()) + 0.5, atol=1e-5)
+
```
[the remark this answers is not in the document]
Comment (gabrieldemarmiesse) [reply tests/keras_contrib/layers/test_capsule.py]: This is pep8 complient, no worries. The build won't (shouldn't) pass if something isn't pep8 complient.

**Comment by gabrieldemarmiesse** [comment]
Thanks @SriRangaTarun for your pull request! It's all nice and pretty now!

**Comment by gabrieldemarmiesse** [comment]
@SriRangaTarun If you have the time one of those days, could you add the squash activation here: https://github.com/keras-team/keras-contrib/tree/master/keras_contrib/activations ? Then you'll be able to update the docstring of capsule.py then with the import.

**Comment by SriRangaTarun** [comment]
@gabrieldemarmiesse, I can probably do that this weekend. Thank you for being so patient and supportive with me !

**Comment by gabrieldemarmiesse** [comment]
@SriRangaTarun no problem and thanks for your contribution!
